From ca49bf6dd309f94ca8b9fdd30cdec596407b5a7f Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 2 Feb 2024 17:05:15 +0100
Subject: [PATCH 0873/1013] evl/net: socket: handle release event

sock_oob_release() is called when the file struct associated to the
socket is closed, which unlike sock_oob_destroy() may block.

Move the calls to evl_release_file() and evl_pass_crossing() to the
release hook, which fixes an issue with sock_oob_destroy() attempting
to block while passing the crossing in a RCU callback context.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/net/socket.h       |  6 ++++--
 kernel/evl/net/packet/packet.c |  8 ++++----
 kernel/evl/net/socket.c        | 35 ++++++++++++++++++++++++++--------
 3 files changed, 35 insertions(+), 14 deletions(-)

diff --git a/include/evl/net/socket.h b/include/evl/net/socket.h
index 1929db05fd4e..9345a3f840ad 100644
--- a/include/evl/net/socket.h
+++ b/include/evl/net/socket.h
@@ -10,7 +10,7 @@
 #include <linux/types.h>
 #include <linux/list.h>
 #include <linux/mutex.h>
-#include <linux/atomic.h>
+#include <linux/refcount.h>
 #include <linux/skbuff.h>
 #include <evl/wait.h>
 #include <evl/file.h>
@@ -27,7 +27,8 @@ struct net_device;
 struct evl_net_proto {
 	int (*attach)(struct evl_socket *esk,
 		struct evl_net_proto *proto, int protocol);
-	void (*detach)(struct evl_socket *esk);
+	void (*release)(struct evl_socket *esk);
+	void (*destroy)(struct evl_socket *esk);
 	int (*bind)(struct evl_socket *esk,
 		struct sockaddr *addr, int len);
 	int (*connect)(struct evl_socket *esk,
@@ -72,6 +73,7 @@ struct evl_socket {
 	struct evl_wait_queue wmem_wait;
 	struct evl_crossing wmem_drain;
 	int protocol;
+	refcount_t refs;	/* release vs destroy */
 	union {
 		struct {
 			int real_ifindex;
diff --git a/kernel/evl/net/packet/packet.c b/kernel/evl/net/packet/packet.c
index e314ede1a000..3f04aedd7d85 100644
--- a/kernel/evl/net/packet/packet.c
+++ b/kernel/evl/net/packet/packet.c
@@ -240,8 +240,8 @@ static int attach_packet_socket(struct evl_socket *esk,
 	return 0;
 }
 
-/* in-band, esk->lock held or socket_release() */
-static void detach_packet_socket(struct evl_socket *esk)
+/* in-band, esk->lock held or __sk_destruct() */
+static void destroy_packet_socket(struct evl_socket *esk)
 {
 	struct evl_net_rxqueue *rxq, *n;
 	unsigned long flags;
@@ -317,7 +317,7 @@ static int bind_packet_socket(struct evl_socket *esk,
 	sll->sll_ifindex = real_ifindex;
 
 	if (esk->protocol != ntohs(sll->sll_protocol)) {
-		detach_packet_socket(esk);
+		destroy_packet_socket(esk);
 		/*
 		 * Since the old binding was dropped, we would not
 		 * receive anything if the new binding fails. This
@@ -660,7 +660,7 @@ static __poll_t poll_packet(struct evl_socket *esk,
 
 static struct evl_net_proto ether_packet_proto = {
 	.attach	= attach_packet_socket,
-	.detach = detach_packet_socket,
+	.destroy = destroy_packet_socket,
 	.bind = bind_packet_socket,
 	.oob_send = send_packet,
 	.oob_poll = poll_packet,
diff --git a/kernel/evl/net/socket.c b/kernel/evl/net/socket.c
index 8b7bda8fcb16..46f80a31f950 100644
--- a/kernel/evl/net/socket.c
+++ b/kernel/evl/net/socket.c
@@ -217,8 +217,8 @@ static struct evl_net_proto *find_oob_proto(int domain, int type, int protocol)
 }
 
 /*
- * In-band call from the common network stack creating a new socket,
- * @sock is already bound to a file. We know the following:
+ * In-band call from the common network stack creating a new BSD
+ * socket, @sock is already bound to a file. We know the following:
  *
  * - the caller wants us either to attach an out-of-band extension to
  *   a common protocol (e.g. AF_PACKET over ethernet), or to set up an
@@ -257,8 +257,10 @@ int sock_oob_attach(struct socket *sock)
 		esk = kzalloc(sizeof(*esk), GFP_KERNEL);
 		if (esk == NULL)
 			return -ENOMEM;
+		refcount_set(&esk->refs, 2); /* release + destroy */
 	} else {
 		esk = (struct evl_socket *)sk;
+		refcount_set(&esk->refs, 1); /* release only */
 	}
 
 	esk->sk = sk;
@@ -306,6 +308,26 @@ int sock_oob_attach(struct socket *sock)
 	return ret;
 }
 
+/*
+ * In-band call from the common network stack releasing a BSD socket,
+ * @sock is still bound to a file, but the network representation
+ * sock->sk might be stale.
+ */
+void sock_oob_release(struct socket *sock)
+{
+	struct evl_socket *esk = evl_sk_from_file(sock->file);
+
+	if (esk->proto->release)
+		esk->proto->release(esk);
+
+	evl_release_file(&esk->efile);
+	/* Wait for the stack to drain in-flight outgoing buffers. */
+	evl_pass_crossing(&esk->wmem_drain);
+
+	if (refcount_dec_and_test(&esk->refs))
+		kfree(esk);
+}
+
 /*
  * In-band call from the common network stack which is about to
  * destruct a socket, releasing all resources attached (@sock is
@@ -315,19 +337,16 @@ void sock_oob_destroy(struct sock *sk)
 {
 	struct evl_socket *esk = evl_sk(sk);
 
-	evl_release_file(&esk->efile);
-	/* Wait for the stack to drain in-flight outgoing buffers. */
-	evl_pass_crossing(&esk->wmem_drain);
 	/* We are detaching, so rmem_count can be left out of sync. */
 	evl_net_free_skb_list(&esk->input);
 
 	evl_destroy_wait(&esk->input_wait);
 	evl_destroy_wait(&esk->wmem_wait);
 
-	if (esk->proto->detach)
-		esk->proto->detach(esk);
+	if (esk->proto->destroy)
+		esk->proto->destroy(esk);
 
-	if (sk->sk_family != PF_OOB)
+	if (sk->sk_family != PF_OOB && refcount_dec_and_test(&esk->refs))
 		kfree(esk);	/* meaning sk != esk. */
 
 	sk->oob_data = NULL;
-- 
2.30.2

